In [1]:
import numpy as np
import pandas as pd

# Load raw data

In [2]:
%%time
train_values = pd.read_csv('sustainable-industry-rinse-over-run-training-set-values.csv',
                           sep=';',
                           index_col=0,
                           parse_dates=['timestamp'])
train_values.head()

C:\Users\ID20178\Anaconda3\lib\site-packages\numpy\lib\arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


Wall time: 5min 56s


object_id               phase                 timestamp pipeline  \
process_id                                                                     
20775             941         final_rinse 2018-03-28 13:32:40+07:00       L9   
20775             941         final_rinse 2018-03-28 13:36:56+07:00       L9   
20775             941         final_rinse 2018-03-28 13:37:20+07:00       L9   
20776             932                acid 2018-04-04 20:26:48+07:00       L8   
20782             113  intermediate_rinse 2018-04-15 08:35:18+07:00       L1   

            supply_flow  supply_pressure  return_temperature  \
process_id                                                     
20775       32392.94100         0.137370           70.518660   
20775       28023.72500         0.244575           19.231047   
20775          57.87037         0.048394           11.476418   
20776       22424.76800        -0.110894           49.048756   
20782          37.61574         0.839627           79.528360   

            return_conductivity  return_turbidity  return_flow  ...  \
process_id                                                      ...   
20775                 41.482605          0.354456   25057.8710  ...   
20775                  0.314453          1.302083   29864.7290  ...   
20775                  0.271579          0.737847   27830.2230  ...   
20776                  0.468332          0.220631   35850.6950  ...   
20782                 41.981667         16.106047    2688.0786  ...   

            tank_temperature_pre_rinse  tank_temperature_caustic  \
process_id                                                         
20775                        27.882668                  83.43461   
20775                        27.759691                  82.36039   
20775                        27.759691                  82.38209   
20776                        27.376303                  82.52315   
20782                        32.259117                  82.72569   

            tank_temperature_acid  tank_concentration_caustic  \
process_id                                                      
20775                   72.276474                   45.153120   
20775                   72.276474                   45.156685   
20775                   72.276474                   45.142090   
20776                   72.576675                   45.195724   
20782                   72.554980                   45.727770   

            tank_concentration_acid  tank_lsh_caustic  tank_lsh_acid  \
process_id                                                             
20775                     44.309620             False              0   
20775                     44.309620             False              0   
20775                     44.309620             False              0   
20776                     44.645237             False              0   
20782                     44.491592             False              0   

            tank_lsh_clean_water  tank_lsh_pre_rinse  target_time_period  
process_id                                                                
20775                      False                   0               False  
20775                      False                   0                True  
20775                      False                   0                True  
20776                      False                   0               False  
20782                      False                   0               False  

[5 rows x 34 columns]

In [3]:
%%time
test_values = pd.read_csv('sustainable-industry-rinse-over-run-test-set-values.csv',
                          sep=';',
                          index_col=0,
                          parse_dates=['timestamp'])
test_values.head()

C:\Users\ID20178\Anaconda3\lib\site-packages\numpy\lib\arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


Wall time: 1min 48s


object_id      phase                 timestamp pipeline  \
process_id                                                            
25031             965    caustic 2018-05-23 16:49:07+07:00      L11   
25031             965    caustic 2018-05-23 16:53:23+07:00      L11   
25034             112    caustic 2018-05-04 00:14:11+07:00       L1   
25039             438  pre_rinse 2018-05-01 00:43:15+07:00       L4   
25039             438    caustic 2018-05-01 00:48:31+07:00       L4   

            supply_flow  supply_pressure  return_temperature  \
process_id                                                     
25031        30213.7580        -0.036892           37.547020   
25031        29953.3420         0.304688           72.012440   
25034         4019.0974         1.309462           50.354458   
25039        15219.9080         1.046224           24.587673   
25039        29998.5530         2.011285           27.423323   

            return_conductivity  return_turbidity  return_flow  ...  \
process_id                                                      ...   
25031                 33.056580          1.486545   21598.3090  ...   
25031                 46.266640          1.182726   33901.1880  ...   
25034                 45.017784          2.658420    4855.3237  ...   
25039                  0.189274          7.794415   13364.4390  ...   
25039                  1.377655          2.025463   38823.7850  ...   

            tank_temperature_pre_rinse  tank_temperature_caustic  \
process_id                                                         
25031                        29.304108                  82.83782   
25031                        29.343895                  81.70934   
25034                        31.405527                  82.94994   
25039                        30.475983                  82.45442   
25039                        30.190250                  82.41464   

            tank_temperature_acid  tank_concentration_caustic  \
process_id                                                      
25031                   72.855180                   45.246970   
25031                   72.710500                   45.035686   
25034                   72.941986                   45.266890   
25039                   73.017940                   45.272274   
25039                   73.017940                   45.315514   

            tank_concentration_acid  tank_lsh_caustic  tank_lsh_acid  \
process_id                                                             
25031                     46.539764             False            0.0   
25031                     44.923610             False            0.0   
25034                     44.618195             False            0.0   
25039                     43.103653             False            0.0   
25039                     43.103653             False            0.0   

            tank_lsh_clean_water  tank_lsh_pre_rinse  target_time_period  
process_id                                                                
25031                      False                 0.0               False  
25031                      False                 0.0               False  
25034                      False                 0.0               False  
25039                      False                 0.0               False  
25039                      False                 0.0               False  

[5 rows x 34 columns]

In [4]:
train_labels = pd.read_csv('sustainable-industry-rinse-over-run-training-labels.csv',
                           sep=';',
                           index_col=0)
train_labels.head()

final_rinse_total_turbidity_liter
process_id                                   
20008                            4.100847e+06
20047                            3.263581e+05
20076                            1.888934e+06
20096                            6.065643e+05
20120                            5.075102e+05

# Drop final rinse

In [5]:
train_view = train_values[train_values['phase'] != 'final_rinse']

This drops processes that only had a final rinse. We drop them from target values for consistency:

In [6]:
train_labels = train_labels.drop(list(set(train_values.index.unique()).difference(set(train_view.index.unique()))))

# Split train data

In [7]:
np.random.seed(123)

In [8]:
train_target = train_labels[['final_rinse_total_turbidity_liter']].sample(frac=0.7)
test_target  = train_labels[['final_rinse_total_turbidity_liter']].drop(train_target.index)

In [9]:
train_timeseries = train_view[train_view.index.isin(train_target.index)]
test_timeseries  = train_view[train_view.index.isin(test_target.index)]

# Truncate

In [10]:
def count_phases(data):
    phase_count = data.groupby('process_id')['phase'].nunique().value_counts().sort_index().rename('count')
    phase_count = (phase_count.rename('pct') / phase_count.sum()).mul(100).round(1)
    return phase_count

In [11]:
count_phases(train_view)

1     4.8
2    20.3
3     0.4
4    74.4
Name: pct, dtype: float64

In [12]:
count_phases(test_values)

1    14.7
2    39.9
3    22.8
4    22.6
Name: pct, dtype: float64

## Truncation method

We randomly select **51.8%** of the 74.4% processes with 4 phases, then randomly:
* select **9.9%** and truncate them to 1 phase;
* select **19.6%** and truncate them to 2 phases;
* select **22.4%** and truncate them to 3 phases.


In [13]:
trunc_params = {
    1: 0.099,  # truncate to 1 phase
    2: 0.196,
    3: 0.224,
}
phases = ['pre_rinse', 'caustic', 'intermediate_rinse', 'acid']

In [14]:
def truncate(data):
    phase_count = data.groupby('process_id')['phase'].nunique()
    nb_processes = len(phase_count)  # percentages are expressed in terms of the total population
    sample_pool = phase_count[phase_count == 4]
    for truncate, pct in trunc_params.items():
        count = round(pct * nb_processes)
        processes = sample_pool.sample(count).index.tolist()
        data = data[~(data.index.isin(processes) & data['phase'].isin(phases[truncate:]))]
        # drop selected processes so they are not selected again
        sample_pool = sample_pool[~sample_pool.index.isin(processes)]
    return data

In [15]:
train_ts_truncated = truncate(train_timeseries)
test_ts_truncated = truncate(test_timeseries)

## Visual check against hidden test set

In [16]:
def drop_consecutive_duplicates(s):
    return ' + '.join(s[s.shift() != s])

def recipe_count(data):
    return data.sort_values(by=['phase']).groupby('process_id')['phase'].apply(drop_consecutive_duplicates).value_counts().rename('count')

In [17]:
pd.concat([recipe_count(test_values).rename('hidden_test'), recipe_count(test_ts_truncated).rename('our_test')], axis=1, sort=False).fillna(0).astype(int)

hidden_test  our_test
caustic + pre_rinse                                     1182       600
acid + caustic + intermediate_rinse + pre_rinse          671       337
caustic + intermediate_rinse + pre_rinse                 670       336
pre_rinse                                                292       149
acid                                                     122        64
caustic                                                   23         9
acid + caustic + intermediate_rinse                        5         6
caustic + intermediate_rinse                               2         0

In [18]:
pd.concat([recipe_count(test_values).rename('hidden_test'), recipe_count(train_ts_truncated).rename('our_train')], axis=1, sort=False).fillna(0).astype(int)

hidden_test  our_train
caustic + pre_rinse                                     1182       1398
acid + caustic + intermediate_rinse + pre_rinse          671        791
caustic + intermediate_rinse + pre_rinse                 670        785
pre_rinse                                                292        350
acid                                                     122        135
caustic                                                   23         29
acid + caustic + intermediate_rinse                        5         16
caustic + intermediate_rinse                               2          0

Looks good - we can ignore low-count recipes.

# Sanity checks

Did we drop data points or columns?

In [19]:
assert len(train_view) == len(train_timeseries) + len(test_timeseries)

In [20]:
assert train_view.shape[1] == train_timeseries.shape[1] == test_timeseries.shape[1]

Have we lost any processes?

In [21]:
assert train_view.index.nunique() == train_ts_truncated.index.nunique() + test_ts_truncated.index.nunique()

Did we bleed data?

In [22]:
assert not set(train_ts_truncated.index.unique()).intersection(test_ts_truncated.index.unique())

Do we have target values for all processes?

In [23]:
assert not set(train_ts_truncated.index.unique()).difference(set(train_target.index))

In [24]:
assert not set(test_ts_truncated.index.unique()).difference(set(test_target.index))

# Save data

In [25]:
# we sort indices to make matching of features and targets easier
train_target = train_target.sort_index()
test_target  =  test_target.sort_index()

In [26]:
train_target.to_csv('train_target.csv')
test_target.to_csv('test_target.csv')

train_ts_truncated.to_csv('train_ts_truncated.csv')
test_ts_truncated.to_csv('test_ts_truncated.csv')

In [27]:
# save to csv without any change
test_values.to_csv('test_values.csv')